### PR TITLE
optimized immutable arrays

### PR DESCRIPTION
**Pull request by KrisVandermotten**
Passing an `ImmutableArray<T>` as an `IEnumerable<T>` creates several issues:

- the value is boxed
- the efficient `foreach` implementation is not used
- the optimized extension methods are not used
- the `Length` property cannot be accessed as easily

This PR fixes several of those issues.

In addition:

- Some of the affected methods were made `static`, where possible.
- Some collections were initialized using the length of the immutable array as the initial capacity, where appropriate.
- A redundant `.ToArray()` call was removed.

**Risk**

All methods are private methods. Risk should be low.

**Performance impact**

The changes are meant to improve the performance of the affected methods.

### CODE REVIEW

File: src/Scripting/Core/Hosting/CommandLine/CommandLineRunner.cs [hunk 1]
```diff
@@ -348,14 +348,12 @@ private void DisplayHelpText()
...
-        private void DisplayDiagnostics(IEnumerable<Diagnostic> diagnostics)
+        private void DisplayDiagnostics(ImmutableArray<Diagnostic> diagnostics)
         {
             const int MaxDisplayCount = 5;
 
-            var errorsAndWarnings = diagnostics.ToArray();
-
             // by severity, then by location
-            var ordered = errorsAndWarnings.OrderBy((d1, d2) =>
+            var ordered = diagnostics.OrderBy((d1, d2) =>
```
Comment (sharwell) [inline src/Scripting/Core/Hosting/CommandLine/CommandLineRunner.cs]: :memo: I verified that this is still lazily calculated

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler and @dotnet/roslyn-analysis for reviewing the parts of this that impact your code

**Comment by sharwell** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest please

**Comment by heejaechang** [comment]
👍 

..

is this to reduce boxing? do we really need to do even this kind of optimization? feels like.. hard to sustain..

**Comment by jaredpar** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by jaredpar** [comment]
Looks like a flaky failure in Roslyn.VisualStudio.IntegrationTests.VisualBasic.BasicExpressionEvaluator.EvaluateInvalidExpressions

https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_debug_vs-integration_prtest/5974/testReport/junit/Roslyn.VisualStudio.IntegrationTests.VisualBasic/BasicExpressionEvaluator/EvaluateInvalidExpressions/

@cston is this known?

**Comment by cston** [comment]
The test failure is https://github.com/dotnet/roslyn/issues/20711.

**Comment by KrisVandermotten** [comment]
It was a different failure on the first run, one I could not reproduce locally. Don't know where this one comes from either.

**Comment by 333fred** [comment]
That failure is https://github.com/dotnet/roslyn/issues/15497.

**Comment by 333fred** [comment]
@dotnet-bot retest 
windows_debug_vs-integration_prtest please

**Comment by jaredpar** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by jaredpar** [comment]
The test runner crashed in the previous failed run.

**Comment by CyrusNajmabadi** [comment]
Something has gone terribly wrong here.

**Comment by KrisVandermotten** [comment]
@CyrusNajmabadi Indeed, and I don't know how to fix it. Any ideas?

**Comment by khyperia** [comment]
@KrisVandermotten It seems as though your latest commit ("rebased" https://github.com/dotnet/roslyn/commit/1aef6f9f90f5c49c0b0a685c05a2cc21ab1daf66 ) went wrong - it has two parent commits, https://github.com/dotnet/roslyn/commit/43795d7f4c54bfd70146647a4f4358df80199236 and https://github.com/dotnet/roslyn/commit/93024bed8454811d8ba15aad13f19c1c3496ecde , which is somehow two versions of the previous commit ("optimized immutable arrays"). I think this is probably pulling in duplicates of the commits on master. I would suggest trying to rebase again (you will probably have to force-push to your branch), or perhaps use a hammer and `git reset --soft master` and re-commiting your changes.

**Comment by KrisVandermotten** [comment]
I tried fixing this, but did not succeed. I see no other option but to start from scratch. I will close this PR.

**Pull request closed without merging**